src/resource/bindless.rs

```diff
@@ -0,0 +1,153 @@
+use std::sync::{Arc, Mutex};
+
+use anyhow::Result;
+
+use ash::vk;
+
+use crate::{pipeline::set_layout::DescriptorSetLayoutCreateInfo, util::cache::Resource};
+
+static MAX_BINDLESS_COUNT: u32 = 4096;
+
+pub trait Bindlessable {
+    fn desc_type() -> vk::DescriptorType;
+    fn res_binding(binding: u32) -> vk::DescriptorSetLayoutBinding {
+        vk::DescriptorSetLayoutBinding {
+            binding,
+            descriptor_type: Self::desc_type(),
+            descriptor_count: MAX_BINDLESS_COUNT,
+            stage_flags: vk::ShaderStageFlags::ALL,
+            p_immutable_samplers: std::ptr::null(),
+        }
+    }
+    fn descriptor_info(&self) -> vk::DescriptorImageInfo;
+}
+
+impl Bindlessable for crate::image::Image {
+    fn desc_type() -> vk::DescriptorType {
+        vk::DescriptorType::STORAGE_IMAGE
+    }
+
+    fn descriptor_info(&self) -> vk::DescriptorImageInfo {
+        todo!()
+    }
+}
+
+impl Bindlessable for (crate::image::ImgView, crate::sampler::Sampler) {
+    fn desc_type() -> vk::DescriptorType {
+        vk::DescriptorType::COMBINED_IMAGE_SAMPLER
+    }
+
+    fn descriptor_info(&self) -> vk::DescriptorImageInfo {
+        vk::DescriptorImageInfo {
+            sampler: unsafe { self.1.handle() },
+            image_view: unsafe { self.0.handle() },
+            image_layout: vk::ImageLayout::SHADER_READ_ONLY_OPTIMAL, //TODO don't hardcode
```

Comment:
This hardcoding could be eliminated by adding an optional field to the type for `CombinedImageSampler` above, and default to `ShaderReadOnlyOptimal` when its not set.